nomad/eval_endpoint.go

```diff
@@ -363,15 +363,45 @@ func (e *Eval) List(args *structs.EvalListRequest,
 			}
 
 			var evals []*structs.Evaluation
+			var itemCount int32
+			var lastToken string
+			foundToken := args.QueryOptions.LastToken == ""
 			for {
 				raw := iter.Next()
 				if raw == nil {
+					lastToken = ""
 					break
 				}
 				eval := raw.(*structs.Evaluation)
+
+				// note: using a gomemdb.FilterIterator doesn't get us
+				// anything here because we'd need all this same logic
+				// in the FilterFunc, as well as an extra type cast
+				if args.FilterJobID != "" && args.FilterJobID != eval.JobID {
+					continue
+				}
+				if args.FilterEvalStatus != "" && args.FilterEvalStatus != eval.Status {
+					continue
+				}
+				if !foundToken {
+					if eval.ID == args.QueryOptions.LastToken {
+						foundToken = true
+					}
+					continue
+				}
+				itemCount++
+				if args.QueryOptions.PerPage != 0 && itemCount > args.QueryOptions.PerPage {
```

Comment:
Yeah this has gotten hairy... I think I'm going to pull the page tracking logic out into its own object so that we can just call `.Next(eval.ID)` and get back an enum that tells us whether to skip, include, or break. That'll clean up the logic in this function and then be reusable across any other paginated RPCs we want.

Reply:
Ok, I've pulled that out as its own component in `nomad/paginator.go`. That made me realize we had a bug in the last token tracking for nil objects, too! See what you think.